rust/query/typeql_define.rs

```diff
@@ -34,23 +39,34 @@ pub struct TypeQLDefine {
 }
 
 impl TypeQLDefine {
-    pub fn new(definables: Vec<Pattern>) -> Self {
-        definables.into_iter().fold(TypeQLDefine::default(), |mut define, definable| {
-            match definable {
-                Pattern::Rule(rule) => define.add_rule(rule),
-                Pattern::Variable(Variable::Type(var)) => define.add_definition(var),
-                _ => unreachable!(),
+    pub fn new(definables: Vec<Definable>) -> Self {
+        definables.into_iter().fold(TypeQLDefine::default(), |define, definable| match definable {
+            Definable::RuleDefinition(rule) => define.add_rule(rule),
+            Definable::TypeVariable(var) => define.add_definition(var),
+            Definable::RuleDeclaration(r) => {
+                panic!("{}", INVALID_RULE_WHEN_MISSING_PATTERNS.format(&[&r.to_string()]))
             }
```

Comment:
I'd quite like to split definable and undefinable into separate enums, but that clashes with the way the grammar is formulated currently.